python/pyarrow/tests/test_cuda_numba_interop.py

```diff
@@ -75,6 +75,89 @@ def make_random_buffer(size, target='host', dtype='uint8', ctx=None):
     raise ValueError('invalid target value')
 
 
+@pytest.mark.parametrize("c", range(len(context_choice_ids)),
+                         ids=context_choice_ids)
+@pytest.mark.parametrize("dtype", dtypes, ids=dtypes)
+@pytest.mark.parametrize("size", [0, 1, 8, 1000])
+def test_from_object(c, dtype, size):
+    ctx, nb_ctx = context_choices[c]
+    arr, cbuf = make_random_buffer(size, target='device', dtype=dtype, ctx=ctx)
+
+    # Creating device buffer from numba MemoryObject:
+    mem = cbuf.to_numba()
+    cbuf2 = ctx.buffer_from_object(mem)
+    assert cbuf2.size == cbuf.size
+    arr2 = np.frombuffer(cbuf2.copy_to_host(), dtype=dtype)
+    np.testing.assert_equal(arr, arr2)
+
+    # Creating device buffer from a device buffer
+    cbuf2 = ctx.buffer_from_object(cbuf2)
+    assert cbuf2.size == cbuf.size
+    arr2 = np.frombuffer(cbuf2.copy_to_host(), dtype=dtype)
+    np.testing.assert_equal(arr, arr2)
+
+    # Creating device buffer from numba DeviceNDArray:
+    darr = DeviceNDArray((size,), (np.dtype(dtype).itemsize,), np.dtype(dtype),
+                         gpu_data=mem)
+    cbuf2 = ctx.buffer_from_object(darr)
+    assert cbuf2.size == cbuf.size
+    arr2 = np.frombuffer(cbuf2.copy_to_host(), dtype=dtype)
+    np.testing.assert_equal(arr, arr2)
+
+    # Creating device buffer from a slice of numba DeviceNDArray:
+    if size >= 8:
+        for s in [slice(size//4, None, None),
```

Comment:
It doesn't make much sense to test _all_ those slices. Also, it seems you're not testing a negative stride here. The important additional cases (apart from the nominal case) are: non-contiguous and negative stride (also, both at the same time).

Reply:
Removed some of the slice tests.

Testing negative slices turned out problematic because of a [numba bug](https://github.com/numba/numba/issues/3705) where slicing numba device array with negative step always produces nonsense. Even reversing a 1D device array via `[::-1]` is wrong in numba. 

Note that the returned buffer of `[::-1]` result will be identical to the buffer of original array.

So, it might make more sense to return the buffer view of the base array when applying `buffer_from_object` to an nd array. Reconstructing the original array from this buffer and array shape information will be easier and there is no penalty in memory or cpu usage.
@pitrou What do you think?

Reply:
Btw, testing N-D array to `CudaBuffer` feature using `numba DeviceNDArray` is just a convenience. Otherwise one would have to cook up an implementation of NDArray with device storage.

Reply:
I'm not sure Approach 1 makes a lot of sense, or is useful. But Approach 2 looks error-prone to me (why would I get a view over something bigger than the argument I'm passing?).

So I would suggest Approach 3: raise an error if the array is non-contiguous, otherwise return a view over the exact array. This will make implementation and test simple, and keep options open if we have a better idea later.

Reply:
Approach 3 sounds good to me as well.

Btw, if an array has negative strides then it is non-contiguous.

Reply:
Similarly, a contiguous array that is neither in C nor in Fortran order will be taken by Numpy as non-contiguous:
```python
>>> a = np.arange(12).reshape((2,2,3))
>>> b = as_strided(a, strides=(24, 48, 8))
>>> b.flags                                                                                                                                                        
  C_CONTIGUOUS : False
  F_CONTIGUOUS : False
  OWNDATA : False
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False
  UPDATEIFCOPY : False
```

Reply:
If `A=np.arange(10)` and `B=A[::-1]`, the contiguous buffers of `A` and `B` would be identical.
Would this be useful or confusing for the users?

Also, the `__cuda_array_interface__` does not include the array storage order information.
The contiguity has to be determined based on shape and strides information only. Such a contiguity would be more general than the numpy definition of contiguity.

For example, `A` and `A.T` would have the same buffer. Just looking at the strides of `A.T`, it is not possible to determine if it is C contiguous or not, for instance.

So, I incline to making `buffer_from_object` successful whenever
```
buffer.size == array.itemsize * product(array.shape)
```
will hold. What do you think?

Reply:
Hmm, actually the C contiguous property can be determined from strides when the itemsize is know. But I wouldn't support only C contiguous objects.

Reply:
>  the contiguous buffers of A and B would be identical. Would this be useful or confusing for the users?

Well, if you're accepting `A` as well as `A.T`, then there's no real reason to reject `A[::-1]` ;-) But I don't really think it's important either way, do what is simplest and we can refine if there's some user request for it.

> What do you think?

That seems good enough to me.

Reply:
Done.